network/phonebook.go

```diff
@@ -61,28 +101,76 @@ func (e *phonebookEntries) ReplacePeerList(they []string) {
 
 	// prepare a map of items we'd like to remove.
 	removeItems := make(map[string]bool, 0)
-	for k := range *e {
+	for k := range e.data {
 		removeItems[k] = true
 	}
 
 	for _, addr := range they {
-		if _, has := (*e)[addr]; has {
+		if _, has := e.data[addr]; has {
 			// we already have this. do nothing.
 			delete(removeItems, addr)
 		} else {
 			// we don't have this item. add it.
-			(*e)[addr] = phonebookData{}
+			e.data[addr] = phonebookData{}
 		}
 	}
 
 	// remove items that were missing in they
 	for k := range removeItems {
-		delete((*e), k)
+		delete(e.data, k)
 	}
 }
 
 func (e *phonebookEntries) updateRetryAfter(addr string, retryAfter time.Time) {
-	(*e)[addr] = phonebookData{retryAfter: retryAfter}
+	_, found := e.data[addr]
+	if !found {
+		e.data[addr] = phonebookData{retryAfter: retryAfter, recentConnectionTimes: make([]time.Time, 0)}
+	} else {
+		entry := e.data[addr]
+		entry.retryAfter = retryAfter
+		e.data[addr] = entry
+	}
+}
+
+// waitAndAddConnectionTime will wait to prevent exceeding connectionsRateLimitingCount.
+// Then it will register the next connection time.
+// Will return true if it waited and false otherwise
+func (e *phonebookEntries) waitAndAddConnectionTime(addr string) bool {
+	waited := false
+	_, found := e.data[addr]
+	curTime := time.Now()
+	if !found {
+		// The addr is not in this phonebook.
+		// Will find the addr in a different phonebook.
+		return waited
+	}
+
+	var timeSince time.Duration
+	var numElmtsToRemove int
+	// Remove from recentConnectionTimes the times later than ConnectionsRateLimitingWindowSeconds
+	for numElmtsToRemove < len(e.data[addr].recentConnectionTimes) {
+		timeSince = curTime.Sub((e.data[addr].recentConnectionTimes)[numElmtsToRemove])
+		if timeSince > e.connectionsRateLimitingWindow {
+			numElmtsToRemove++
+		} else {
+			break // break the loop. The rest are earlier than 1 second
+		}
+	}
+	e.popNElements(numElmtsToRemove, addr)
+
+	// If there are max number of connections within the time window, wait
+	numElts := len(e.data[addr].recentConnectionTimes)
+	if uint(numElts) >= e.connectionsRateLimitingCount {
+		// Wait until the earliest time expires
+		time.Sleep(e.connectionsRateLimitingWindow - timeSince)
+		waited = true
+		// Remove it from recentConnectionTimes
+		e.popNElements(1, addr)
+	}
+
+	// Append the time for the next connection request
+	e.appendTimeNow(addr)
```

Comment:
instead of letting this function determine what is "now", re-use the `curTime`

Reply:
Actually, this logic is wrong.

Consider the case where multiple parallel requests are made to the server. For some reason, it takes a long time to connect to the server. The recorded connected times will be much earlier than when the server actually received the requests. 

Then, the client makes more requests, since the recorded times are expired by now. However, this may exceed the limit for the server. 

To correct this, I am separating the wait for the connection from the add time. The add time should be made after the request was sent.

Reply:
I think that the code is correct - but we need to "append" before and "update" once the connection is established.